### PR TITLE
[BEAM-1146] Decrease spark runner startup overhead

### PR DESCRIPTION
**Pull request by aviemzur**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Replace finding all `Source` and `Coder` implementations for serialization registration with wrapper classes.

### CODE REVIEW

**Comment by aviemzur** [comment]
R: @amitsela @kennknowles

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderWrapper.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package org.apache.beam.runners.spark.coders;
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/CoderWrapper.java]: Need license header here.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java [hunk 2]
```diff
@@ -20,59 +20,24 @@
...
-          }));
-    }
-
-    static Iterable<Class<?>> getClasses() {
-      return INSTANCE;
-    }
+  @Override
+  public void registerClasses(Kryo kryo) {
+    kryo.register(MicrobatchSource.class, new JavaSerializer());
+    kryo.register(SourceRDD.Bounded.BoundedSourceWrapper.class, new JavaSerializer());
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: I am not so familiar with kryo registration. Can you register `BoundedSource` without a wrapper, or does it not work for superclasses?

Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: Correct, registering a superclass or interface does not handle implementations or extensions. This is the reason for the previous implementation here which found all descendants using `Reflections` library. By using a wrapper we can reduce performance overhead involved with reflection.

**Comment by aviemzur** [comment]
Looks like there are still a few more usages of coders in aggregators and side input which are not covered yet. I'm looking into it

**Comment by aviemzur** [comment]
After digging deeper into `Spark` serialization code it seems that the issue is simpler than we first thought.
`Spark` serializes closures using Java Serializer, not Kryo. Most of the `Coder` instances in Spark runner are within such closures.
Since `Source` instances are serialized in both bounded and unbounded pipelines, these did require wrappers (`UnboundedSource` was already wrapped in `MicrobatchSource` and `BoundedSource` now has a wrapper named `BoundedSourceWrapper`. The latter might need a better name).
These `Source` instances often have `Coder` members themselves, but the serialization of these is already taken care of by having the `Source` itself registered to be serialized with `JavaSerializer`.

**Comment by dhalperi** [comment]
This looks great to me! I know that @amitsela may be out of touch, and Kenn is on vacation.

Should I merge it, @aviemzur ?

Thanks!!
Dan

**Comment by amitsela** [comment]
@aviemzur looks great!
I'm not sure you need `BoundedSourceWrapper` or registering `NamedAggregators.CombineFunctionState.class` because I think they will also be serialized with `JavaSerializer`.

**Comment by aviemzur** [comment]
@amitsela 

You're right, `RDD` instances are serialized using the `Closure` serializer, which, as mentioned above uses Java serialization. For example see [here](https://github.com/apache/spark/blob/v1.6.3/core/src/main/scala/org/apache/spark/scheduler/ResultTask.scala#L60)

Accumulator updates as well as seen [here](https://github.com/apache/spark/blob/v1.6.3/core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala#L57)

Changed the commit to reflect this, now only registering `MicrobatchSource` to serialize using Java.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java [hunk 3]
```diff
@@ -39,40 +30,7 @@
...
-              return reflections.getSubTypesOf(clazz);
-            }
-          }));
-    }
-
-    static Iterable<Class<?>> getClasses() {
-      return INSTANCE;
-    }
+    // MicrobatchSource is serialized as data and may not be Kryo-serializable.
+    kryo.register(MicrobatchSource.class, new StatelessJavaSerializer());
```
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: 👍

**Comment by amitsela** [comment]
I'll merge.

**Pull request merged**